Mathlib/Lean/Meta/RefinedDiscrTree/Pi.lean

```diff
@@ -0,0 +1,269 @@
+  if h : args.size < 2 then return none else some <$> do
+  let type := args[0]
+  let inst := args[1]
+  etaExpand args type lambdas 3 fun args lambdas _ => do
+    let arg := args[2]
+    reduceAux type inst [arg] lambdas args #[inst] 3 1
+      (lambdaBody? 2 >=> (·.app3? n) >=> fun (_, inst, a) =>
+        -- check for `fun a x => -a x` or `fun a x => (a x)⁻¹`
+        if (a == .app (.bvar 1) (.bvar 0)) then inst else none)
+      (fun
+        | type, [arg] => #[type, none, arg]
+        | _, _ => unreachable! )
+
+/-- Normalize an application if the head is `•` or `+ᵥ`. -/
+private def reduceHActOp (n : Name) (args : Array Expr) (lambdas : List FVarId) :
+    MetaM (Option (List (Array (Option Expr) × List FVarId))) := withDefault do
+  if h : args.size < 5 then return none else
+  let rec instH := match n with
+    | `HVAdd.hVAdd => `instHVAdd
+    | `HSMul.hSMul => `instHSMul
+    | _ => unreachable!
+  let some (α, type, inst) := args[3].app3? instH | return none
+  etaExpand args type lambdas 6 fun args lambdas _ => do
+    let a := args[4]
+    let arg := args[5]
+    reduceAux type inst [arg] lambdas args #[inst, a] 6 2
+      (lambdaBody? 3 >=> app6? n >=> fun (inst, a, b) =>
+        -- check for `fun a b x => a • b x` or `fun a b x => a +ᵥ b x`
+        if (a == .bvar 2) && b == .app (.bvar 1) (.bvar 0) then inst else none)
+      (fun
+        | type, [arg] => #[α, type, none, none, a, arg]
+        | _, _ => unreachable!)
+
+/-- Normalize an application if the head is `^`. -/
+private def reduceHPow (args : Array Expr) (lambdas : List FVarId) :
+    MetaM (Option (List (Array (Option Expr) × List FVarId))) := withDefault do
+  if h : args.size < 6 then return none else
+  let some (type, β, inst) := args[3].app3? ``instHPow | return none
+  let arg := args[4]
+  let exp := args[5]
+  reduceAux type inst [arg] lambdas args #[inst, exp] 6 2
+      (lambdaBody? 3 >=> app6? ``HPow.hPow >=> fun (inst, a, b) =>
+        -- check for `fun a b x => a x ^ b`
+        if (a == .app (.bvar 2) (.bvar 0) && b == .bvar 1) then inst else none)
+    (fun
+      | type, [arg] => #[type, β, none, none, arg, exp]
+      | _, _ => unreachable!)
+
+
+/-- Normalize an application if the head is `+`, `*`, `-` or `/`. -/
+private def reduceHBinOp (n : Name) (args : Array Expr) (lambdas : List FVarId):
+    MetaM (Option (List (Array (Option Expr) × List FVarId))) := withDefault do
+  if h : args.size < 4 then return none else do
+  let rec instH := match n with
+    | ``HAdd.hAdd => ``instHAdd
+    | ``HMul.hMul => ``instHMul
+    | ``HSub.hSub => ``instHSub
+    | ``HDiv.hDiv => ``instHDiv
+    | _ => unreachable!
+  let some (type, inst) := args[3].app2? instH | return none
+  etaExpand args type lambdas 6 fun args lambdas _ => do
+    let lhs := args[4]
+    let rhs := args[5]
+    reduceAux type inst [lhs, rhs] lambdas args #[inst] 6 2
+      (lambdaBody? 3 >=> app6? n >=> fun (inst, l, r) =>
+        -- check for the form `fun a b x => a x + b x`
```

Comment:
Is this cheaper than doing an `isDefEq` check against the canonical instances, as would happen with `~q()` matching? It certainly seems more fragile this way.

Reply:
It is cheaper indeed, but this code won't run too many times, so it's probably fine to use isDefEq instead.